add content about nt-dml for insert and update (#11449)
pingcap/docs: non-transactional-dml.md

A non-transactional DML statement is a DML statement split into multiple SQL statements (which is, multiple batches) to be executed in sequence. It enhances the performance and ease of use in batch data processing at the expense of transactional atomicity and isolation.

> Usually, large transactions that consume too much memory need to be split into multiple SQL statements to bypass the transaction size limit. Non-transactional DML statements integrate this feature into the TiDB kernel to achieve the same effect. You can understand the effect of non-transactional DML statements by splitting SQL statements. The `DRY RUN` syntax can be used to preview the split statements.

Reviewer: ```suggestion
Usually, memory-consuming large transactions need to be split into multiple SQL statements to bypass the transaction size limit. Non-transactional DML statements integrate this process into the TiDB kernel to achieve the same effect. You can understand the effect of non-transactional DML statements by splitting SQL statements. The `DRY RUN` syntax can be used to preview the split statements.
```

Non-transactional DML statements include `INSERT`, `REPLACE`, `UPDATE`, and `DELETE`. For detailed syntax, see [`BATCH`](/sql-statements/sql-statement-batch.md).

> **Note:**
>
> - A non-transactional DML statement does not guarantee the atomicity and isolation of the statement, and is not equivalent to the original DML statement.

> > - After a DML statement is rewritten into a non-transactional DML statement, you cannot assume that its behavior is consistent with the original one.

Reviewer: ```suggestion
> - After a DML statement is rewritten into a non-transactional DML statement, you cannot assume that its behavior is consistent with that of the original statement.
```

A non-transactional DML statement is a DML statement split into multiple SQL statements (which is, multiple batches) to be executed in sequence. It enhances the performance and ease of use in batch data processing at the expense of transactional atomicity and isolation.

> Usually, memory-consuming transactions need to be split into multiple SQL statements to bypass the transaction size limit. Non-transactional DML statements integrate this process into the TiDB kernel to achieve the same effect. You can understand the effect of non-transactional DML statements by splitting SQL statements. The `DRY RUN` syntax can be used to preview the split statements.

Reviewer: ```suggestion
Usually, memory-consuming transactions need to be split into multiple SQL statements to bypass the transaction size limit. Non-transactional DML statements integrate this process into the TiDB kernel to achieve the same effect. It is helpful to understand the effect of non-transactional DML statements by splitting SQL statements. The `DRY RUN` syntax can be used to preview the split statements.
```

- The statement does not require atomicity, which permits some rows to be modified and some rows to remain unmodified in the execution result.
- The statement is idempotent, or you are prepared to retry on a part of the data according to the error message. If the system variables are set to `tidb_redact_log = 1` and `tidb_nontransactional_ignore_error = 1`, this statement must be idempotent. Otherwise, when the statement partially fails, the failed part cannot be accurately located.

> - The data to be operated on has no other concurrent writes, which means it is not updated by other statements at the same time. Otherwise, unexpected results such as missing writing, wrong writing, and modifying the same line multiple times might occur.

Reviewer: ```suggestion
- The data to be operated on has no other concurrent writes, which means it is not updated by other statements at the same time. Otherwise, unexpected results such as missing writes, wrong writes, and modifying the same line multiple times might occur.
```

- The statement does not modify the data to be read by the statement itself. Otherwise, the following batch will read the data written by the previous batch and easily causes unexpected results.

>     - The split column should not be updated in the statement. For example, for a non-transactional `UPDATE` statement, the split SQL statements are executed in sequence. The modification of the previous batch is read by the next batch after it is committed, which causes the same line of data to be modified multiple times.

Author: Done in https://github.com/pingcap/docs/pull/11449/commits/73374fc2054e609b012374a836ad2511cbd397a3

>     - The shard column should not be updated in the statement. For example, for a non-transactional `UPDATE` statement, the split SQL statements are executed in sequence. The modification of the previous batch is read by the next batch after it is committed, which causes the same line of data to be modified multiple times.

Reviewer: ```suggestion
    - The dividing column should not be updated in the statement. For example, for a non-transactional `UPDATE` statement, the split SQL statements are executed in sequence. The modification of the previous batch is read by the next batch after the previous batch is committed, which causes the same line of data to be modified multiple times.
```

>     - The split column should not be updated in the statement. For example, for a non-transactional `UPDATE` statement, the split SQL statements are executed in sequence. The modification of the previous batch is read by the next batch after it is committed, which causes the same line of data to be modified multiple times.

Author: > It is called shard column

It is `dividing column` in [Non-Transactional DML Statements](https://docs.pingcap.com/tidb/stable/non-transactional-dml#parameter-description)

>     - The shard column should not be used as a Join key. For example, the following example uses the shard column `test.t.id` as a Join key, which causes a non-transactional `UPDATE` statement to modify the same line multiple times:

Reviewer: ```suggestion
    - The dividing column should not be used as a Join key. For example, the following example uses the dividing column `test.t.id` as a Join key, which causes a non-transactional `UPDATE` statement to modify the same line multiple times:
```

- The data to be operated on has no other concurrent writes, which means it is not updated by other statements at the same time. Otherwise, unexpected results such as missing writes, wrong writes, and modifying the same line multiple times might occur.
- The statement does not modify the data to be read by the statement itself. Otherwise, the following batch will read the data written by the previous batch and easily causes unexpected results.

    - The shard column should not be modified when selecting from and modifying the same table within an non-transactional `INSERT INTO ... SELECT` statement. Otherwise, multiple batches might read the same row and insert data multiple times:
        - It is not recommended to use `BATCH ON test.t.id LIMIT 10000 INSERT INTO t SELECT id+1, value FROM t;`.
        - It is recommended to use `BATCH ON test.t.id LIMIT 10000 INSERT INTO t SELECT id, value FROM t;`.
        - If the shard column `id` has the `AUTO_INCREMENT` attribute, it is recommended to use `BATCH ON test.t.id LIMIT 10000 INSERT INTO t(value) SELECT value FROM t;`.

>     - The shard column should not be updated in the non-transactional `UPDATE`, `INSERT ... ON DUPLICATE KEY UPDATE`, or `REPLACE INTO` statement:

Reviewer: ```suggestion
    - Avoid updating the shard column in the non-transactional `UPDATE`, `INSERT ... ON DUPLICATE KEY UPDATE`, or `REPLACE INTO` statement:
```

1 row in set
```

The following example describes how to use multiple table joins. First, create table `t2` and insert data:

```sql
CREATE TABLE t2(id int, v int, key(id));
INSERT INTO t2 VALUES (1,1), (3,3), (5,5);
```

> Then, update the data of table `t2` by joining table `t` and `t2`. Note that the shard column needs to be specified with the complete database name, table name, and column name (`test.t.id`):

Reviewer: ```suggestion
Then, update the data of table `t2` by joining table `t` and `t2`. Note that the dividing column needs to be specified with the complete database name, table name, and column name (`test.t.id`):
```

1 row in set
```

The following example describes how to use multiple table joins. First, create table `t2` and insert data:

```sql
CREATE TABLE t2(id int, v int, key(id));
INSERT INTO t2 VALUES (1,1), (3,3), (5,5);
```

> Then, update the data of table `t2` by joining table `t` and `t2`. Note that the shard column needs to be specified with the complete database name, table name, and column name (`test.t.id`):

Reviewer: ```suggestion
Then, update the data of table `t2` by joining table `t` and `t2`. Note that you need to specify the shard column along with the complete database name, table name, and column name (`test.t.id`):
```

To use a non-transactional DML statement, the following steps are recommended:

1. Select an appropriate [dividing column](#parameter-description). Integer or string types are recommended.
2. Add `DRY RUN QUERY` to the non-transactional DML statement, execute the query manually, and confirm whether the data range affected by the DML statement is roughly correct.

> 3. Add `DRY RUN` to the non-transactional DML statement, execute the query manually, and check the split statements and the execution plans. You need to pay attention to whether the split statement can read the result written by the previous statement, which might cause abnormal phenomena. You also need to pay attention to the index selection efficiency.

Reviewer: ```suggestion
3. Add `DRY RUN` to the non-transactional DML statement, execute the query manually, and check the split statements and the execution plans. You need to pay attention to whether the split statement can read the result written by the previous statement, which might cause an anomaly. You also need to pay attention to index selection efficiency.
```
